File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/design_linked_list.rb
class Node
    attr_accessor :val, :next
    def initialize(val)
        @val = val
        @next = nil
    end
end

class MyLinkedList

=begin
    Initialize your data structure here.
=end
    attr_accessor :head, :length
    def initialize()
        @length = 0
        @head = nil
    end


=begin
    Get the value of the index-th node in the linked list. If the index is invalid, return -1.
    :type index: Integer
    :rtype: Integer
=end
    def get(index)
        case
            when index < 0 then return -1
            when index == 0 then return @head.val
            when index >= @length then return -1
        else
            current = @head
            for i in 0..(index-1)
                current = current.next
            end
            return -1 if current.nil?
            return current.val
        end
    end


=begin
    Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
    :type val: Integer
    :rtype: Void
=end
    def add_at_head(val)
        first = Node.new(val)
        first.next = @head
        @head = first
        @length += 1
    end


=begin
    Append a node of value val to the last element of the linked list.
    :type val: Integer
    :rtype: Void
=end
    def add_at_tail(val)
        last = Node.new(val)
        current = @head
        if @length == 0 then
            @head = last
            @length += 1
        else
            while !current.next.nil?
                current = current.next
            end
            current.next = last
            @length += 1
        end
    end


=begin
    Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
    :type index: Integer
    :type val: Integer
    :rtype: Void
=end
    def add_at_index(index, val)
        if index < 0 || index > @length
            return
        elsif index == 0
            add_at_head(val)
        elsif index == @length
            add_at_tail(val)
        else
            insert = Node.new(val)
            prev = @head
            for i in 0..(index-2)
                prev = prev.next
            end
            succ = prev.next
            prev.next = insert
            insert.next = succ
            @length += 1
        end
    end


=begin
    Delete the index-th node in the linked list, if the index is valid.
    :type index: Integer
    :rtype: Void
=end
    def delete_at_index(index)
        if index < 0 || index >= @length
            return
        elsif index == 0
            @head = @head.next
            @length -= 1
        else
            current = @head
            for i in 0..(index-2)
                current = current.next
            end
            return if current.next.nil?
            current.next = current.next.next
            @length += 1
        end
    end

        
end

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList.new()
# param_1 = obj.get(index)
# obj.add_at_head(val)
# obj.add_at_tail(val)
# obj.add_at_index(index, val)
# obj.delete_at_index(index)

File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/design_linked_list.go
type MyLinkedNode struct {
    val int
    next *MyLinkedNode
    prev *MyLinkedNode
}


type MyLinkedList struct {
    head *MyLinkedNode
    tail *MyLinkedNode
}


/** Initialize your data structure here. */
func Constructor() MyLinkedList {
    return MyLinkedList{nil, nil}
}


func (this *MyLinkedList) getNode(index int) (*MyLinkedNode, int) {
    if index < 0 {
        return nil, -1
    } else {
        var i = 0
        var node = this.head
        for node != nil && i < index {
            node = node.next
            i++
        }
        return node, i
    }
}


/** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
func (this *MyLinkedList) Get(index int) int {
    var node, _ = this.getNode(index)
    if node == nil {
        return -1
    }
    return node.val
}


/** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
func (this *MyLinkedList) AddAtHead(val int)  {
    if this.head == nil {
        this.head = &MyLinkedNode{val: val, next: nil, prev: nil}
        this.tail = this.head
    } else {
        var node = &MyLinkedNode{val: val, next: this.head, prev: nil}
        this.head.prev = node
        this.head = node
    }
}


/** Append a node of value val to the last element of the linked list. */
func (this *MyLinkedList) AddAtTail(val int)  {
    if this.tail == nil {
        this.AddAtHead(val)
    } else {
        var node = &MyLinkedNode{val: val, next: nil, prev: this.tail}
        this.tail.next = node
        this.tail = node
    }
}


/** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
func (this *MyLinkedList) AddAtIndex(index int, val int)  {
    if index == 0 {
        this.AddAtHead(val)
    } else {
        var node, i = this.getNode(index)
        if node == nil {
            if i == index {
                this.AddAtTail(val)
            }
        } else {
            var insert = &MyLinkedNode{val: val, next: node, prev: node.prev}
            node.prev.next = insert
            node.prev = insert
        }
    }
}


/** Delete the index-th node in the linked list, if the index is valid. */
func (this *MyLinkedList) DeleteAtIndex(index int)  {
    var node, _ = this.getNode(index)
    if node == nil {
        return
    }
    if node.next != nil {
        node.next.prev = node.prev
    }
    if node.prev != nil {
        node.prev.next = node.next
    }
    if node == this.head {
        this.head = node.next
    }
    if node == this.tail {
        this.tail = node.prev
    }
}


/**
 * Your MyLinkedList object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */
